src/solvers/dgsem_tree/indicators_2d.jl

```diff
@@ -24,6 +24,100 @@ function create_cache(typ::Type{IndicatorHennemannGassner}, mesh, equations::Abs
   create_cache(typ, equations, dg.basis)
 end
 
+# Overload indicator when ShallowWaterEquations2D is used to apply full FV method on cells
+# containing some "dry" LGL nodes. That is, if an element is partially "wet" then it becomes a
+# full FV element.
+function (indicator_hg::IndicatorHennemannGassner)(u::AbstractArray{<:Any,4},
+                                                   mesh, equations::ShallowWaterEquations2D, 
+                                                   dg::DGSEM, cache; kwargs...)
+  @unpack alpha_max, alpha_min, alpha_smooth, variable = indicator_hg
+  @unpack alpha, alpha_tmp, indicator_threaded, modal_threaded, modal_tmp1_threaded = indicator_hg.cache
+  # TODO: Taal refactor, when to `resize!` stuff changed possibly by AMR?
+  #       Shall we implement `resize!(semi::AbstractSemidiscretization, new_size)`
+  #       or just `resize!` whenever we call the relevant methods as we do now?
+  resize!(alpha, nelements(dg, cache))
+  if alpha_smooth
+    resize!(alpha_tmp, nelements(dg, cache))
+  end
+
+  # magic parameters
+  threshold = 0.5 * 10^(-1.8 * (nnodes(dg))^0.25)
+  parameter_s = log((1 - 0.0001)/0.0001)
+
+  # If h at one LGL node is lower than threshold_wet, set alpha[element]=1 via indicator_wet
+  # in order to apply a full FV method in the partially wet cell and guarantee the well-balanced property.
+  #
+  # Determination of hard coded threshold: Showed good results in many numerical experiments. Idea is to
+  # gain more stability when computing the velocity on (nearly) dry cells which
+  # could be counteracted by division of conservative variables, e.g., v2 / v1 = hv / h.
+  # Here, the impact of the threshold on the number of cells being updated with FV is not that
+  # significant. However, its impact on the stability is very significant.
+  # The value can be seen as a trade-off between accuracy and stability.
+  # Well-balancedness of the scheme on partially wet cells with hydrostatic reconstruction 
+  # can only be proven for the FV method (see Chen and Noelle).
+  # Therefore we set alpha to one regardless of its given maximum value.
+  #    
+  threshold_wet = 1e-4
+
+  @threaded for element in eachelement(dg, cache)
+    indicator  = indicator_threaded[Threads.threadid()]
+    modal      = modal_threaded[Threads.threadid()]
+    modal_tmp1 = modal_tmp1_threaded[Threads.threadid()]
+
+    # (Re-)set dummy variable for alpha_dry
+    indicator_wet = 1.0
+
+    # Calculate indicator variables at Gauss-Lobatto nodes
+    for j in eachnode(dg), i in eachnode(dg)
+      u_local = get_node_vars(u, equations, dg, i, j, element)
+      h, _, _, _ = u_local
+      indicator_wet = min(indicator_wet, Int32(h > threshold_wet))
```

Comment:
This seems to be a convoluted way of writing
```suggestion
      if h <= threshold_wet
        indicator_wet = 0
      end
```
Isn't it? If you do it in the proposed way, you'd have to set `indicator_wet = 1` above as well, to avoid a type instability.

Reply:
@ranocha @sloede I changed the logic of this clipping to your recommended way with `if-clauses` everywhere. I also removed my old idea of avoiding limiting the bottom variable and deleted the new method `set_node_vars`